python/lsst/daf/butler/registry/databases/sqlite.py

```diff
@@ -0,0 +1,217 @@
+
+    def __str__(self) -> str:
+        return f"SQLite3@{self.filename}"
+
+    def _convertFieldSpec(self, table: str, spec: ddl.FieldSpec, metadata: sqlalchemy.MetaData,
+                          **kwds) -> sqlalchemy.schema.Column:
+        if spec.autoincrement:
+            if not spec.primaryKey:
+                raise RuntimeError(f"Autoincrement field {table}.{spec.name} that is not a "
+                                   f"primary key is not supported.")
+            if spec.dtype != sqlalchemy.Integer:
+                # SQLite's autoincrement is really limited; it only works if
+                # the column type is exactly "INTEGER".  But it also doesn't
+                # doesn't care about the distinctions between different integer
+                # types, so it's safe to change it.
+                spec = copy.copy(spec)
+                spec.dtype = sqlalchemy.Integer
+        return super()._convertFieldSpec(table, spec, metadata, **kwds)
+
+    def _convertTableSpec(self, name: str, spec: ddl.TableSpec, metadata: sqlalchemy.MetaData,
+                          **kwds) -> sqlalchemy.schema.Table:
+        primaryKeyFieldNames = set(field.name for field in spec.fields if field.primaryKey)
+        autoincrFieldNames = set(field.name for field in spec.fields if field.autoincrement)
+        if len(autoincrFieldNames) > 1:
+            raise RuntimeError("At most one autoincrement field per table is allowed.")
+        if len(primaryKeyFieldNames) > 1 and len(autoincrFieldNames) > 0:
+            # SQLite's default rowid-based autoincrement doesn't work if the
+            # field is just one field in a compound primary key.
+            # As a workaround, we create an extra table with just one column
+            # that we'll insert into to generate those IDs.
+            autoincrFieldName, = autoincrFieldNames
+            self._autoincr[name] = _AutoincrementWorkaround(
+                table=self._convertTableSpec(f"_autoinc_{name}", _AUTOINCR_TABLE_SPEC, metadata, **kwds),
+                column=autoincrFieldName
+            )
+        return super()._convertTableSpec(name, spec, metadata, **kwds)
+
+    def insert(self, table: sqlalchemy.schema.Table, *rows: dict, returnIds: bool = False,
+               ) -> Optional[List[int]]:
+        autoincr = self._autoincr.get(table.name)
+        if autoincr is not None:
+            # This table has a compound primary key that includes an
+            # autoincrement.  That doesn't work natively in SQLite, so we
+            # insert into this single-column table and use those IDs.
+            if not rows:
+                return [] if returnIds else None
+            with self.transaction():
+                newRows = []
+                ids = []
+                for row in rows:
+                    newRow = row.copy()
+                    id = self._connection.execute(autoincr.table.insert()).inserted_primary_key[0]
+                    newRow[autoincr.column] = id
```

Comment:
Will look at it later today or tomorrow.

Reply:
It looks OK to me though I think there is a potential for confusion here. I myself do not understand very well how autoincrement is supposed to work in case of composite key, maybe it's a general problem and not an issue with our implementation. As an example suppose we have a table with `PK=(auto, origin)` and we insert a row `(origin=1)` first which generates `auto=1`. Next we insert explicit `(auto=1, origin=2)` which I think is a valid operation and should work in general, but this will cause issues with our implementation because autoincr table already has `id=1`. I do not know how to solve this, maybe removing a row from autoincr table before insert (or checking that id is not there, which should be cheaper) would work?

Reply:
The code on the new commit (59e7189b05239c067e5cdf32465de1dfde9c2867) tries to handle exactly this case, by inserting into the autoincr table only when `origin == self.origin`.  That means your example will work if and only if `self.origin==1`, but we prohibit the caller from inserting e.g. just (`origin=2)` without an explicit `auto` value.  And that limitation is okay for all of the use cases I have identified.

Of course, that's for SQLite, where this is all manually implemented.  I don't know what the behavior is supposed to be for other databases, but PostgreSQL and Oracle pass the tests I have so far, and those (now) cover this case.  My guess is that they will just skip generating some `auto` values that would have been safe, but that's fine.